### PR TITLE
NIFI-5629 GetFile vast listing performance

### PR DESCRIPTION
**Pull request by adyoun2**
See related PR. When the number of files to pickup becomes a certain size, the OS can become slow performing the listing, so re-use this listing multiple times.

### CODE REVIEW

**Comment by adyoun2** [comment]
How do I trigger a re-build without dummy commits? The build failure appears to be a timeout unrelated to the changes.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-documentation/src/test/java/org/apache/nifi/documentation/DocGeneratorTest.java [hunk 1]
```diff
@@ -72,7 +74,12 @@ public void testProcessorLoadsNarResources() throws IOException, ClassNotFoundEx
     }
 
     private NiFiProperties loadSpecifiedProperties(final String propertiesFile, final String key, final String value) {
-        String file = DocGeneratorTest.class.getResource(propertiesFile).getFile();
+        String file = propertiesFile;
```
Comment (patricker) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-documentation/src/test/java/org/apache/nifi/documentation/DocGeneratorTest.java]: Was this included on accident? Does not seem relevant to your ticket.

File: nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java [hunk 2]
```diff
@@ -458,6 +458,9 @@ public MockFlowFile importFrom(final Path path, final boolean keepSourceFile, Fl
 
         newFlowFile.setData(baos.toByteArray());
         newFlowFile = putAttribute(newFlowFile, CoreAttributes.FILENAME.key(), path.getFileName().toString());
+        if (!keepSourceFile) {
```
Comment (patricker) [inline nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java]: Was this included on accident? Does not seem relevant to your ticket.

File: nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/NiFiProperties.java [hunk 3]
```diff
@@ -1422,10 +1424,15 @@ public String getDefaultBackPressureDataSizeThreshold() {
     public static NiFiProperties createBasicNiFiProperties(final String propertiesFilePath, final Map<String, String> additionalProperties) {
         final Map<String, String> addProps = (additionalProperties == null) ? Collections.EMPTY_MAP : additionalProperties;
         final Properties properties = new Properties();
-        final String nfPropertiesFilePath = (propertiesFilePath == null)
+        String nfPropertiesFilePath = (propertiesFilePath == null)
                 ? System.getProperty(NiFiProperties.PROPERTIES_FILE_PATH)
                 : propertiesFilePath;
         if (nfPropertiesFilePath != null) {
+            try {
```
Comment (patricker) [inline nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/NiFiProperties.java]: Was this included on accident? Does not seem relevant to your ticket.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestGetFile.java [hunk 4]
```diff
@@ -151,6 +151,90 @@ public void testFilePickedUp() throws IOException {
...
+        successFiles.get(1).assertContentEquals("Hello, World!".getBytes("UTF-8"));
+        successFiles.get(2).assertContentEquals("Hello, World!".getBytes("UTF-8"));
+
+        final String path = successFiles.get(2).getAttribute("path");
+        assertEquals("/", path);
+        final String absolutePath = successFiles.get(2).getAttribute(CoreAttributes.ABSOLUTE_PATH.key());
+        assertEquals(absTargetPathStr, absolutePath);
+    }
+
+    @Test
```
Comment (patricker) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestGetFile.java]: I still have more testing to do, but so far this test looks good at demonstrating the issue. When i run this with the unmodified GetFile it runs out steam around 9993 files.

**Comment by patricker** [comment]
@adyoun2 Also, to retrigger the build, I found online that you can close and re-open your PR (no need for a new PR, just re-open). That _should_ re-trigger the checks.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestGetFile.java (hunk 4 above)
Comment (patricker) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestGetFile.java]: On further testing of this test case, weird results. Does this test work for you? I get a mix of Actual counts. I ran this test on the full code and get Actual 9996. I took out the limit, and got an actual of 9999. Not really sure what's going on.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-documentation/src/test/java/org/apache/nifi/documentation/DocGeneratorTest.java (hunk 1 above)
Comment (adyoun2) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-documentation/src/test/java/org/apache/nifi/documentation/DocGeneratorTest.java]: I certainly don't remember making this change. I suspect it changed on master since I forked, but I'll check.

Comment (adyoun2) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-documentation/src/test/java/org/apache/nifi/documentation/DocGeneratorTest.java]: Sorry, it was relevant, to get it working in my local environment, where someone had put spaces in the path, %20 in resource paths needed converting back to ' '

File: nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java (hunk 2 above)
Comment (adyoun2) [reply nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java]: It is necessary, in that the mock was ignoring the attribute, so when tested with multiple files, behaviour was non-deterministic, as it could quite easily pick up the same file again.

File: nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/NiFiProperties.java (hunk 3 above)
Comment (adyoun2) [reply nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/NiFiProperties.java]: it was relevant, to get it working in my local environment, where someone had put spaces in the path, %20 in resource paths needed converting back to ' '

**Comment by patricker** [comment]
@adyoun2 Looks like you are having some issues with your Git, I've been there many times... But I'm sure we can get it working.

Make sure that after you rebase your Github fork you push the changes back to Github. Right now Github thinks you made a lot of extra changes to your branch that you didn't (thus the ~500 changed files). Also, when you squash, don't squash your changes in with another users, just squash your own changes.

**Comment by adyoun2** [comment]
Doh. How do I resolve the situation now, can I just back out the squashed commit?

**Comment by adyoun2** [comment]
Backed out the squashed commit

File: nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/NiFiProperties.java (hunk 3 above)
Comment (patricker) [reply nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/NiFiProperties.java]: This sounds like it could be a legitimate bug, but it would need to be submitted separately as it's not related to fixing `GetFile`. I run on Windows, but I don't have spaces in the path, maybe switch to a new path with no spaces for now?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetFile.java [hunk 5]
```diff
@@ -270,6 +280,9 @@ private FileFilter createFileFilter(final ProcessContext context) {
         return new FileFilter() {
             @Override
             public boolean accept(final File file) {
+                if (file.isDirectory()) {
+                    return false;
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetFile.java]: does filtering work differently with the streams?  I ask because we want to be able to support the same depth/recursive searches and this filter seems like it would stop at dir level.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetFile.java [hunk 6]
```diff
@@ -308,32 +321,45 @@ public boolean accept(final File file) {
...
+                }
+                
+                if (!directory.exists()) {
+                    return Collections.emptySet();
+                }
+                
+                Stream<Path> listStream;
                 if (recurseSubdirectories) {
-                    queue.addAll(performListing(child, filter, recurseSubdirectories));
+                    listStream = Files.walk(p);
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetFile.java]: @markap14 hadn't we switched to using the streaming api previously and backed this out?  If so what was the problem?  Or maybe that was ListFile?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetFile.java (hunk 5 above)
Comment (adyoun2) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetFile.java]: If you look at the performListing, the filter is applied after the walk, with no feedback mechanism, so the additional filter just prevents Directories being returned as entities for GetFile to process.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetFile.java (hunk 6 above)
Comment (markap14) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetFile.java]: That was ListFile. IIRC, we did re-introduce it. The issue the first time was that it was not following links, but that was due to needing to explicitly send in a flag when performing the listing to follow links - the NIO.2 API didn't follow symlinks by default, while the IO API did.

Comment (adyoun2) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetFile.java]: I assume you'll want the same flags set here?

Comment (markap14) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetFile.java]: Yes, I believe so. Most users are going to expect it to follow symlinks. Especially since that's the current behavior. We don't want to break backward compatibility.

**Comment by adyoun2** [comment]
Could someone take a look at this?

Cheers

**Comment by adyoun2** [comment]
Could someone take a look at this?

Cheers

**Comment by joewitt** [comment]
closing due to inactivity
